servers/rendering/renderer_rd/renderer_canvas_render_rd.h

```diff
@@ -335,6 +336,110 @@ class RendererCanvasRenderRD : public RendererCanvasRender {
 
 	//state that does not vary across rendering all items
 
+	struct InstanceData {
+		float world[6];
+		uint32_t flags;
+		uint32_t specular_shininess;
+		union {
+			//rect
+			struct {
+				float modulation[4];
+				union {
+					float msdf[4];
+					float ninepatch_margins[4];
+				};
+				float dst_rect[4];
+				float src_rect[4];
+				float pad[2];
+			};
+			//primitive
+			struct {
+				float points[6]; // vec2 points[3]
+				float uvs[6]; // vec2 points[3]
+				uint32_t colors[6]; // colors encoded as half
+			};
+		};
+		float color_texture_pixel_size[2];
+		uint32_t lights[4];
+	};
+
+	struct PushConstant {
+		uint32_t base_instance_index;
+		uint32_t pad1;
+		uint32_t pad2;
+		uint32_t pad3;
+	};
+
+	struct TextureStateKey {
```

Comment:
This key determines when the texture state for the rendering pipeline must change, and reduces the number of branches for testing individual state changes when batching up draw calls.